Stratis.Bitcoin/BlockPulling/BlockPullerBehavior.cs

```diff
@@ -92,27 +102,40 @@ public override object Clone()
         /// <param name="message">Received message.</param>
         private void Node_MessageReceived(Node node, IncomingMessage message)
         {
+            // TODO: https://github.com/stratisproject/StratisBitcoinFullNode/issues/292
+            this.logger.LogTrace($"[{this.GetHashCode():x}] ({nameof(node.Peer.Endpoint)}:'{node.Peer.Endpoint}')");
+
             message.Message.IfPayloadIs<BlockPayload>((block) =>
             {
-                block.Object.Header.CacheHashes();
-                this.QualityScore = Math.Min(BlockPuller.MaxQualityScore, this.QualityScore + 1);
-
-                foreach (Transaction tx in block.Object.Transactions)
-                    tx.CacheHashes();
-
+                // There are two pullers for each peer connection and each is having its own puller behavior.
+                // Both these behaviors get notification from the node when it receives a message,
+                // even if the origin of the message was from the other puller behavior.
+                // Therefore we first make a quick check whether this puller behavior was the one 
+                // who should deal with this block.
                 uint256 blockHash = block.Object.Header.GetHash();
-                DownloadedBlock downloadedBlock = new DownloadedBlock()
+                if (this.puller.CheckBlockTaskAssignment(this, blockHash))
```

Comment:
I am actually strongly convinced here that this is better. The situation you are describing takes about 0.1 milliseconds. We have usually 8 peers max, but even if we had 100 peers, the chance of running to each other within that 0.1 millisecond is nothing more than plain zero :-)